nixos/modules/security/pam.nix

```diff
@@ -369,233 +409,272 @@ let
+        }) ++ (optional (cfg.showMotd && config.users.motd != null) {
+          control = "optional"; path = "${pkgs.pam}/lib/security/pam_motd.so"; arguments = ["motd=${motd}"];
+        }) ++ (optional (cfg.enableAppArmor && config.security.apparmor.enable) {
+          control = "optional"; path = "${pkgs.apparmor-pam}/lib/security/pam_apparmor.so"; arguments = ["order=user,group,default" "debug"];
+        }) ++ (optional (cfg.enableKwallet) ({
+          control = "optional"; path = "${pkgs.plasma5.kwallet-pam}/lib/security/pam_kwallet5.so";
+          arguments = "kwalletd=${pkgs.kdeFrameworks.kwallet.bin}/bin/kwalletd5";
+        })) ++ (optional (cfg.enableGnomeKeyring) {
+          control = "optional"; path = "${pkgs.gnome3.gnome-keyring}/lib/security/pam_gnome_keyring.so"; arguments = ["auto_start"];
+        }) ++ (optional cfg.gnupg.enable {
+          control = "optional"; path = "${pkgs.pam_gnupg}/lib/security/pam_gnupg.so";
+          arguments = optional cfg.gnupg.noAutostart "no-autostart";
+        }) ++ (optional (config.virtualisation.lxc.lxcfs.enable) {
+          control = "optional"; path = "${pkgs.lxc}/lib/security/pam_cgfs.so"; arguments = [ "-c all" ];
+        });
+      };
     };
-
   };
 
-
   inherit (pkgs) pam_krb5 pam_ccreds;
 
   use_ldap = (config.users.ldap.enable && config.users.ldap.loginPam);
   pam_ldap = if config.users.ldap.daemon.enable then pkgs.nss_pam_ldapd else pkgs.pam_ldap;
 
   # Create a limits.conf(5) file.
-  makeLimitsConf = limits:
-    pkgs.writeText "limits.conf"
-       (concatMapStrings ({ domain, type, item, value }:
-         "${domain} ${type} ${item} ${toString value}\n")
-         limits);
+  makeLimitsConf = limits: pkgs.writeText "limits.conf" (
+    concatMapStrings ({ domain, type, item, value }:
+      "${domain} ${type} ${item} ${toString value}\n"
+    ) limits
+  );
 
   motd = pkgs.writeText "motd" config.users.motd;
 
-  makePAMService = name: service:
-    { name = "pam.d/${name}";
-      value.source = pkgs.writeText "${name}.pam" service.text;
-    };
-
-in
-
-{
+  makePAMService = name: service: let
+    argumentToString = value: if hasInfix " " value then "[${escape [ "]" ] value}]" else value;
+    controlToString = value:
+      if isString value then value
+      else concatStringsSep " " (mapAttrsToList (name: value: "${name}=${toString value}") value);
+    entryToString = moduleType: { control, path, arguments}: (concatStringsSep " " (
+      [ moduleType (controlToString control) path ] ++ (map argumentToString arguments)
+    ));
+  in nameValuePair ("pam.d/${name}") {
+    text = concatStringsSep "\n" (flatten (mapAttrsToList (type: map (entryToString type)) service.entries));
+  };
+in {
 
-  imports = [
-    (mkRenamedOptionModule [ "security" "pam" "enableU2F" ] [ "security" "pam" "u2f" "enable" ])
-  ];
+  imports = [(mkRenamedOptionModule [ "security" "pam" "enableU2F" ] [ "security" "pam" "u2f" "enable" ])];
```

Comment:
I think we can remove this for 21.03 (or 21.05!), people have had long enough to change it, and if we're changing the whole external interface, it isn't that relevant anymore